server/src/main/java/org/elasticsearch/plugins/CircuitMemoryBreakerPlugin.java

```diff
@@ -0,0 +1,53 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.plugins;
+
+import org.elasticsearch.common.settings.ClusterSettings;
+import org.elasticsearch.indices.breaker.BreakerSettings;
+import org.elasticsearch.indices.breaker.CircuitBreakerService;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.function.Consumer;
+
+/**
+ * An extension point for {@link Plugin} implementations to add custom circuit breakers
+ */
+public interface CircuitMemoryBreakerPlugin {
+
+    /**
+     * Returns additional circuit breaker settings added by this plugin.
+     *
+     * This each of the {@link BreakerSettings} are passed to the configured {@link CircuitBreakerService}.
+     * The service will create a new breaker according to the provided settings and overall environment.
+     *
+     */
+    default List<BreakerSettings> getCircuitBreakers() {
+        return Collections.emptyList();
+    }
+
+    /**
+     * If the custom {@link BreakerSettings} have dynamic settings.
+     * A settings update consumer can be added with this method.
+     * Once the new settings are constructed, pass them to the updatedBreakerSettingsListener.
+     */
+    default void addDynamicBreakerUpdates(ClusterSettings settings, Consumer<BreakerSettings> updatedBreakerSettingsListener) {
```

Comment:
All of our plugin apis should be "pull" based, where we extract from the plugin what will be added to ES. This seems to do the opposite. It's unclear what from my reading of this PR what this method actual does, since there are no example or test implementations, but based on the one use inside Node, it seems like this is used as a way to callback to actually add the breaker to the CircuitBreakerService. I think the root of the problem is CircuitBreakerService should not have a registerBreaker method. Instead, Node should collect all the breaker settings, and pass them to the service's ctor, as we don't need the ability to add these dynamically.

Additionally, the BreakerSettings concept seems like it needs a little work to make this all truly extendable. Currently it takes in all the values, but the plugin must read these values on its own, and the `getCircuitBreakers()` method above then relies on the plugin having Settings from its ctor, which cannot be relied on. Even if we passed the real Settings to `getCircuitBreakers()`, it seems opposite of what we want. IMO any new breakers that are added should be configurable in the same way our existing breakers are configured, ie through the same yaml settings with the new breaker name substituted.  This would mean replacing the hardcoded settings that exist for the builtin breakers with affix settings, so that they can be used with any breaker.

Reply:
> It's unclear from my reading of this PR what this method actually does, since there are no example or test implementations, but based on the one use inside Node, it seems like this is used as a way to callback to actually add the breaker to the CircuitBreakerService

It is a way to hook into dynamic setting updates for the new circuit breakers. If this is possible by have a settings listener on a suffix, that would definitely be better. 

> IMO any new breakers that are added should be configurable in the same way our existing breakers are configured, ie through the same yaml settings with the new breaker name substituted.

Effectively, new circuit breakers would be added by plugins declaring new settings that match the appropriate affix?

Reply:
@rjernst I have updated the setting handling. Now new breakers are affix settings. The initial breaker settings are still provided by the plugin, but facilities are in place to make that as seamless as possible.

Still an open question around making the current circuit breakers affix settings and deprecating their current setting values.

Also, I opted to NOT make `durability` configurable. This seems like something a user should NEVER be able to change (even at start up time).

Reply:
It still seems like there is more boiler plate than there should be. What is the plugin actually providing? The setting parsing/registration should be possible within core, where plugins never need to do anything to handle it. Instead, plugins can just provide the breaker name, and core registers the appropriate settings/update handlers. All the plugin needs is a reference to the CircuiteBreaker so it can make calls to add/remove allocations. Currently this is done by exposing the CircuitBreakerService, which has a get method based on the breaker name. This isn't the worst for now, but the registerBreaker method should go away. It was added in a time when we were still fully utilizing guice, instead of explicit construction and pull based plugin apis that we have now.

Reply:
> All the plugin needs is a reference to the CircuitBreaker so it can make calls to add/remove allocations. 

This is not true. The underlying references change when there are setting changes. The plugin won't lose the data it was accounting, but it would not see setting updates until a reboot. Which effectively makes plugin breaker settings static.

> The setting parsing/registration should be possible within core, where plugins never need to do anything to handle it. Instead, plugins can just provide the breaker name, and core registers the appropriate settings/update handlers. 

How can different plugins provide their appropriate setting defaults? To my knowledge, affix settings have global defaults and different plugins might want to provide their own sane defaults.

Reply:
> The underlying references change when there are setting changes. 

This seems pretty bad. There are ways to make the writing use locking but the read side, which is more common, not slow down. Also, after looking at the register method again, I've noticed the comment noting registering the same name replaces the existing breaker. This is leniency which I think makes having this available to plugins error prone. What if two plugins provide the same breaker name? Then the plugin registration order matters, and one  silently replaces the other. This is yet another reason I think we need to statically define the breakers at node startup, but still have them available.

> How can different plugins provide their appropriate setting defaults?

This is a good point and something I hadn't considered. There are ways around this though. For example, in ScriptService we now use affix settings for context specific cache values, where the defaults for those are defined in the contexts themselves (ie like circuit breakers types here). It means the default defined in the setting is ignored and we apply the defaults dynamically based on whether the setting exists, rather than calling `setting.get` always.

Reply:
> There are ways to make the writing use locking but the read side, which is more common, not slow down. 

Keeping the reference the same across threads will require overhead & limit to be `volatile` on read no? How else will threads ever see the updates? I think this will have a non-negligible performance hit. We might get around this hit by making method local copies of the volatile fields so they are only read once per method call.

Reply:
What we have done in other places is wrap the state into a single internal object (eg look at XPackLicenseState.Status). Updates to the setttings construct a new object, and flip all the values at once through a single `volatile`.

> I think this will have a non-negligible performance hit

This is the pattern we have used in numerous places in Elasticsearch. I know of no other place where we poll a service for the current instance of an object on every call. Having services/objects update themselves through cluster state listeners and volatile state is very common, and not a performance hit in any cases I know of.

Reply:
@rjernst volatile reads are 3x as slow as non-volatile. But, given the other changes, I think it will be OK (we won't need concurrent hash map any longer).

I am going to work on implementing this (initializing via ctor, volatile settings, etc.) and test performance with JMH. Will get back once done. 

Thanks so much for your thoughts and time on this!